### PR TITLE
Implement auto redirect

### PR DESCRIPTION
**Pull request by Daltz333**
TODO

- [x] When writing to file, always use forward slashes
- [x] When writing to file, always surround in quotes
- [x] Write tests

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -31,6 +31,11 @@ To check that deleted/renamed files in your git repo are in your redirects,
 1. Make sure `rediraffe_branch` and `rediraffe_redirects` are set in conf.py.
 2. Run the `rediraffecheckdiff` builder.
 
+### Auto Redirect builder
+For convenience, an auto redirect builder has been added. Ensure that `rediraffe_auto_redirect = True` is in your conf.py. Then simply run the `rediraffewritediff` builder.
```
Comment (TheTripleV) [inline README.md:35]:
```suggestion
For convenience, the auto redirect builder can be used to automatically add renamed files to your redirects file. Ensure that `rediraffe_auto_redirect = True` is in your conf.py. Then simply run the `rediraffewritediff` builder.
```

File: README.md [hunk 2]
```diff
@@ -31,6 +31,11 @@ To check that deleted/renamed files in your git repo are in your redirects,
...
+
+Note: The auto redirect builder only works with a configuration file.
+
```
Comment (TheTripleV) [inline README.md:38]:
```suggestion
Note: Deleted files cannot be added to your redirects file automatically.
```

File: sphinxext/rediraffe.py [hunk 3]
```diff
@@ -320,19 +319,57 @@ def abs_path_in_src_dir_w_src_suffix(filename: str) -> Union[Path, None]:
                 logger.error(err_msg)
                 self.app.statuscode = 1
 
+        file = open(path, "a")
```
Comment (TheTripleV) [inline sphinxext/rediraffe.py:322]: Use a context manager for files.

File: tests/test_builder.py [hunk 4]
```diff
@@ -4,7 +4,7 @@
 
 @pytest.fixture(scope="module")
 def rootdir():
-    return path(__file__).parent.abspath() / "roots/builder"
+    return path(__file__).parent.abspath().joinpath("roots").joinpath("builder")
```
Comment (TheTripleV) [inline tests/test_builder.py:7]:
```suggestion
    return path(__file__).parent.abspath() / "roots" / "builder"
```
^ Does that work? It reads better.

File: sphinxext/rediraffe.py [hunk 5]
```diff
@@ -320,19 +319,57 @@ def abs_path_in_src_dir_w_src_suffix(filename: str) -> Union[Path, None]:
...
+                            )
+                            logger.error(err_msg)
+                            self.app.statuscode = 1
+                else:
+                    err_msg = (
+                        f"{red('(broken)')} {renamed_file} was deleted but is not redirected!"
+                        f" Hint: This file was renamed to {hint_to} with a similarity of {perc}%."
+                    )
+                    logger.error(err_msg)
+                    self.app.statuscode = 1
```
Comment (TheTripleV) [inline sphinxext/rediraffe.py:370]: ew consecutive code duplication. I'm wanna play around with this section (324-370). It's getting really verbose.

File: tests/test_builder.py (hunk 4 above)
Comment (Daltz333) [reply tests/test_builder.py:7]: Does the forward slash create an actual forward slash, or use the os.sep(). Joinp path uses is.sep()

Comment (TheTripleV) [reply tests/test_builder.py:7]: `/` just calls joinpath.

**Comment by TheTripleV** [comment]
My PR to this branch addresses the comments to `rediraffe.py`.

A few upkeep things:
1. Could you add a few more tests, covering more complex setups & `rediraffe_auto_redirect_perc`? The one test case was passing but was relying on an ambiguous variable.
2. Can you further separate the checkdiff and writediff tests? (change test names or wrap in separate classes / files). I had created the tests when there was only 1 builder.

**Comment by TheTripleV** [comment]
Depends on #9. This PR wraps filenames in quotes.

**Comment by TheTripleV** [comment]
Closes #4.

File: sphinxext/rediraffe.py [hunk 6]
```diff
@@ -320,20 +320,35 @@ def abs_path_in_src_dir_w_src_suffix(filename: str) -> Union[Path, None]:
...
+                if renamed_file in absolute_redirects:
+                    logger.info(
+                        f"renamed file {renamed_file} redirects to {absolute_redirects[renamed_file]}."
+                    )
+                    continue
+
+                if self.name == "rediraffewritediff":
+                    if perc >= self.app.config.rediraffe_auto_redirect_perc:
+                        rel_rename_from = f'"{relpath(renamed_file, src_path)}"'
+                        rel_rename_to = f'"{relpath(hint_to, src_path)}"'
```
Comment (TheTripleV) [inline sphinxext/rediraffe.py:337]:
```suggestion
                        rel_rename_from = f'"{str(PurePosixPath(renamed_file.relative_to(src_path)))}"'
                        rel_rename_to = f'"{str(PurePosixPath(hint_to.relative_to(src_path)))}"'
```
works on Windows. You need to add `PurePosixPath` to the pathlib imports.

**Pull request merged**